crates/c-api/src/component.rs

```diff
@@ -0,0 +1,659 @@
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_component_instance_get_func(
+    instance: &wasmtime_component_instance_t,
+    context: CStoreContextMut<'_>,
+    name: *const u8,
+    len: usize,
+    item: &mut *mut wasmtime_component_func_t,
+) -> bool {
+    let name = crate::slice_from_raw_parts(name, len);
+    let name = match std::str::from_utf8(name) {
+        Ok(name) => name,
+        Err(_) => return false,
+    };
+    let func = instance.instance.get_func(context, name);
+    if let Some(func) = func {
+        *item = Box::into_raw(Box::new(wasmtime_component_func_t { func }));
+    }
+    func.is_some()
+}
+
+#[repr(transparent)]
+pub struct wasmtime_component_func_t {
+    func: Func,
+}
+
+fn call_func(
+    func: &wasmtime_component_func_t,
+    mut context: CStoreContextMut<'_>,
+    raw_params: &[wasmtime_component_val_t],
+    raw_results: &mut [wasmtime_component_val_t],
+) -> Result<()> {
+    let params = func
+        .func
+        .params(context.as_context())
+        .iter()
+        .zip(raw_params.iter())
+        .map(|(ty, v)| v.clone().into_val(ty))
+        .collect::<Result<Vec<_>>>()?;
+    let mut results = vec![Val::Bool(false); raw_results.len()];
+    func.func.call(context.as_context_mut(), &params, &mut results)?;
+    func.func.post_return(context)?;
+    for (i, r) in results.iter().enumerate() {
+        raw_results[i] = r.try_into()?;
+    }
+    Ok(())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_component_func_call(
+    func: &wasmtime_component_func_t,
+    context: CStoreContextMut<'_>,
+    params: *const wasmtime_component_val_t,
+    params_len: usize,
+    results: *mut wasmtime_component_val_t,
+    results_len: usize,
+    out_trap: &mut *mut wasm_trap_t,
+) -> Option<Box<wasmtime_error_t>> {
+    let raw_params = crate::slice_from_raw_parts(params, params_len);
+    let mut raw_results = crate::slice_from_raw_parts_mut(results, results_len);
+    match call_func(func, context, &raw_params, &mut raw_results) {
+        Ok(_) => None,
+        Err(e) => handle_call_error(e, out_trap),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        wasmtime_component_val_flags_set, wasmtime_component_val_flags_t,
+        wasmtime_component_val_flags_test,
+    };
+
+    #[test]
+    fn bit_fiddling() {
+        let mut flags: wasmtime_component_val_flags_t = Vec::new().into();
+        wasmtime_component_val_flags_set(&mut flags, 1, true);
+        assert!(wasmtime_component_val_flags_test(&flags, 1));
+        assert!(!wasmtime_component_val_flags_test(&flags, 0));
+        wasmtime_component_val_flags_set(&mut flags, 260, true);
+        assert!(wasmtime_component_val_flags_test(&flags, 260));
+        assert!(!wasmtime_component_val_flags_test(&flags, 261));
+        assert!(!wasmtime_component_val_flags_test(&flags, 259));
+        assert!(wasmtime_component_val_flags_test(&flags, 1));
+        assert!(!wasmtime_component_val_flags_test(&flags, 0));
+    }
```

Comment:
When you mentioned not having good tests for the C-API @alexcrichton can you elaborate what is the preferred way of adding those tests? I see maybe 3 options outside of the "examples" that we have today.

1. Add in crate unit tests like here, these would be written in Rust, but using the C-API directly, so would feel like a very C style of Rust.
2. Add in tests to the "integration" style of tests in the `tests` top level director for the C-API, but written in Rust using this crate directly. There is probably a variant of this where we use bindgen to regenerate Rust bindings from the headers to also test the headers are typed correctly.
3. Add new "integration" style of tests written in C (or C++) similar to option 2, but just using the binding headers instead of rust directly.